Subject: [PATCH] add test code
---
Index: compiler/src/main/java/android/databinding/tool/reflection/annotation/ksp/KspAnnotationMethod.kt
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
diff --git a/compiler/src/main/java/android/databinding/tool/reflection/annotation/ksp/KspAnnotationMethod.kt b/compiler/src/main/java/android/databinding/tool/reflection/annotation/ksp/KspAnnotationMethod.kt
--- a/compiler/src/main/java/android/databinding/tool/reflection/annotation/ksp/KspAnnotationMethod.kt	(revision 11a020af7530cf48933a1ef9cf1559ac6bdd8e51)
+++ b/compiler/src/main/java/android/databinding/tool/reflection/annotation/ksp/KspAnnotationMethod.kt	(date 1757470194739)
@@ -93,7 +93,7 @@
     }
 
     override fun getName(): String {
-        return mExecutableElement.simpleName.toString()
+        return mExecutableElement.simpleName.asString()
     }
 
     override fun getReturnType(args: List<ModelClass>): ModelClass {
Index: README.md
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
diff --git a/README.md b/README.md
--- a/README.md	(revision 11a020af7530cf48933a1ef9cf1559ac6bdd8e51)
+++ b/README.md	(date 1757473728224)
@@ -29,3 +29,42 @@
 getAllFields
 ./gradlew :publishLocal
 
+
+3.ERROR: Could not find accessor viewmodel.xxx
+bindingTarget.resolveMultiSetters()
+FieldAccessExpr.resolveType()
+mGetter = resolvedType.findGetterOrField(mName, isStatic)
+
+java.lang.NoClassDefFoundError: android/databinding/Bindable
+
+java.util.NoSuchElementException: No TypeParameter found for index T
+  at com.squareup.kotlinpoet.ksp.TypeParameterResolver$Companion$EMPTY$1.get(TypeParameterResolver.kt:45)
+  at com.squareup.kotlinpoet.ksp.KsTypesKt.toTypeName(KsTypes.kt:76)
+  at com.squareup.kotlinpoet.ksp.KsTypesKt.toTypeName(KsTypes.kt:65)
+  at com.squareup.kotlinpoet.ksp.KsTypesKt.toTypeName$default(KsTypes.kt:63)
+  at android.databinding.tool.reflection.annotation.ksp.KspAnnotationClass.equals(KspAnnotationClass.kt:159)
+  at android.databinding.tool.reflection.annotation.ksp.KspAnnotationClass.isAssignableFrom(KspAnnotationClass.kt:172)
+
+
+java.lang.IllegalArgumentException: couldn't make a guess for
+  at com.squareup.javapoet.Util.checkArgument(Util.java:64)
+  at com.squareup.javapoet.ClassName.bestGuess(ClassName.java:183)
+  at android.databinding.tool.ext.ExtKt.toTypeName(ext.kt:260)
+  at android.databinding.tool.ext.ExtKt.toTypeName(ext.kt:201)
+  at android.databinding.tool.reflection.ModelClass.getTypeName(ModelClass.kt:335)
+  at android.databinding.tool.reflection.ModelClass$isObject$2.invoke(ModelClass.kt:146)
+  at android.databinding.tool.reflection.ModelClass$isObject$2.invoke(ModelClass.kt:145)
+  at kotlin.UnsafeLazyImpl.getValue(Lazy.kt:81)
+  at android.databinding.tool.reflection.ModelClass.isObject(ModelClass.kt:145)
+  at android.databinding.tool.store.SetterStore.calculateConversionPriority(SetterStore.java:1064)
+  at android.databinding.tool.store.SetterStore.isBetterParameter(SetterStore.java:943)
+  at android.databinding.tool.store.SetterStore.getBestSetter(SetterStore.java:803)
+  at android.databinding.tool.store.SetterStore.getSetterCall(SetterStore.java:625)
+  at android.databinding.tool.Binding.resolveSetterCall(Binding.java:140)
+  at android.databinding.tool.Binding.getSetterCall(Binding.java:108)
+  at android.databinding.tool.Binding.injectSafeUnboxing(Binding.java:245)
+  at android.databinding.tool.BindingTarget.injectSafeUnboxing(BindingTarget.java:160)
+  at android.databinding.tool.LayoutBinder.<init>(LayoutBinder.java:156)
+  at android.databinding.tool.DataBinder.<init>(DataBinder.java:61)
+
+
Index: compiler/src/main/java/android/databinding/tool/reflection/annotation/ksp/KspAnnotationClass.kt
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
diff --git a/compiler/src/main/java/android/databinding/tool/reflection/annotation/ksp/KspAnnotationClass.kt b/compiler/src/main/java/android/databinding/tool/reflection/annotation/ksp/KspAnnotationClass.kt
--- a/compiler/src/main/java/android/databinding/tool/reflection/annotation/ksp/KspAnnotationClass.kt	(revision 11a020af7530cf48933a1ef9cf1559ac6bdd8e51)
+++ b/compiler/src/main/java/android/databinding/tool/reflection/annotation/ksp/KspAnnotationClass.kt	(date 1757473141786)
@@ -154,7 +154,8 @@
 
     override fun equals(other: Any?): Boolean {
         if (other is KspAnnotationClass) {
-            return typeMirror.toTypeName() == other.typeMirror.toTypeName()
+            return typeMirror.equals(other.typeMirror)
+//            return typeMirror.toTypeName() == other.typeMirror.toTypeName()
         }
         return false
     }
Index: compiler/src/main/java/android/databinding/tool/reflection/annotation/ksp/KspAnnotationField.kt
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
diff --git a/compiler/src/main/java/android/databinding/tool/reflection/annotation/ksp/KspAnnotationField.kt b/compiler/src/main/java/android/databinding/tool/reflection/annotation/ksp/KspAnnotationField.kt
--- a/compiler/src/main/java/android/databinding/tool/reflection/annotation/ksp/KspAnnotationField.kt	(revision 11a020af7530cf48933a1ef9cf1559ac6bdd8e51)
+++ b/compiler/src/main/java/android/databinding/tool/reflection/annotation/ksp/KspAnnotationField.kt	(date 1757470194647)
@@ -39,7 +39,7 @@
     }
 
     override fun getName(): String {
-        return mField.simpleName.toString()
+        return mField.simpleName.asString()
     }
 
     override fun isPublic(): Boolean {
Index: compilerCommon/src/main/kotlin/android/databinding/tool/BindableCompat.kt
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
diff --git a/compilerCommon/src/main/kotlin/android/databinding/tool/BindableCompat.kt b/compilerCommon/src/main/kotlin/android/databinding/tool/BindableCompat.kt
--- a/compilerCommon/src/main/kotlin/android/databinding/tool/BindableCompat.kt	(revision 11a020af7530cf48933a1ef9cf1559ac6bdd8e51)
+++ b/compilerCommon/src/main/kotlin/android/databinding/tool/BindableCompat.kt	(date 1757472027829)
@@ -32,12 +32,14 @@
     companion object {
         @JvmStatic
         fun extractFrom(element : KSFunctionDeclaration) : BindableCompat? {
-            return extractSupport(element) ?: extractAndroidX(element)
+//            return extractSupport(element) ?: extractAndroidX(element)
+            return extractAndroidX(element)
         }
 
         @JvmStatic
         fun extractFrom(element : KSPropertyDeclaration) : BindableCompat? {
-            return extractSupport(element) ?: extractAndroidX(element)
+//            return extractSupport(element) ?: extractAndroidX(element)
+            return extractAndroidX(element)
         }
 
         @JvmStatic
